### ISSUE TITLE
Implement bracket alignment rule from flake8-hangover

### ISSUE THREAD

**Issue by adamsol** [astral-sh/ruff#23842]
### Summary

Continuation from https://github.com/astral-sh/ruff/issues/21596#issuecomment-4025340193.

The idea is to enforce formatter-like code style using a linter rule. Implementation proposal is in #23809. The upstream rule is [here](https://github.com/deniskrumko/flake8-hangover?tab=readme-ov-file#fhg005-close-bracket-have-different-indentation-with-open-bracket).

@MichaReiser:

> My main concern is formatter compatibility (e.g. hugging parens). 

As far as I'm aware, the rule does not conflict with the formatter.

It flags some additional cases around multiline strings (as noted in the PR), but only because the formatter doesn't enforce a fully consistent style around them (probably because it cannot change indentation in the string content). If the indentation is fixed manually, both the formatter and the rule should be happy.

Bracket hugging is accepted but not forced. It's in the docs and test file only to demonstrate how the rule behaves in this situation.

> We've also haven't added any new formatter lints in a while, so while the rule might be different, I'm not sure it's something we want to add right now.

While I understand the reluctance to maintaining two ways of achieving a similar goal, there are many open issues with the autoformatter, related to code readability, which prevent me and other users from adopting it. For example: #11753, #11886, #12856, #13761, #21925.

This is why it would be useful to have some linter rules that can accomplish a similar goal as the formatter, without having to sacrifice our manual formatting.

**Comment by ntBre** [comment]
To save others a couple of clicks, the rule flags code like this:

```py
# ERROR: Close bracket on line with last parameter not allowed
func(
    123,
    456)
```

suggesting to change it to:

```py
# OK: Close bracket on new line
func(
    123,
    456,
)
```

which is the reason why the other discussion was on overlap with COM812. COM812 doesn't currently flag the first code, which Micha noted [here](https://github.com/astral-sh/ruff/issues/9216#issuecomment-1961688768). If COM812 did flag this and suggested adding the magic trailing comma to yield:

```py
func(
    123,
    456,)
```

the formatter would then give the desired output. But I understand from the issue summary that this rule is meant to operate without running the formatter, so adding this behavior to COM812 wouldn't be sufficient.

https://play.ruff.rs/dbe23a48-5873-4a40-be23-17488c922caf?secondary=Format

**Comment by adamsol** [comment]
Thanks for the summary! Everything is correct, and I'd add that:

* COM812 won't flag this case flagged by HG005 (closing bracket with mismatched indentation):
  ```
  func(
      123,
      456,
    )
  ```
* HG005 won't flag this case flagged by COM812 (missing trailing comma):
  ```
  func(
      123,
      456
  )
  ```

This is why I argued that the rules are independent.
